Tools/speech_recognizer/GSSpeechRecognitionServer.m

```diff
@@ -0,0 +1,146 @@
+   but WITHOUT ANY WARRANTY; without even the implied warranty of
+   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+   Lesser General Public License for more details.
+   
+   You should have received a copy of the GNU Lesser General Public
+   License along with this library; if not, write to the Free
+   Software Foundation, Inc., 51 Franklin Street, Fifth Floor,
+   Boston, MA 02110 USA.
+*/
+
+#import "GSSpeechRecognitionServer.h"
+#import "GSSpeechRecognitionEngine.h"
+#import <Foundation/Foundation.h>
+
+static GSSpeechRecognitionServer *_sharedInstance;
+static int _clients = 0;
+
+@implementation GSSpeechRecognitionServer
+
+/**
+ * Monitor connection...
+ */
++ (void)connectionDied: (NSNotification*)aNotification
+{
+  NSArray *objs = [[aNotification object] localObjects];
+  NSEnumerator *en = [objs objectEnumerator];
+  id o = nil;
+
+  if(_clients > 0)
+    {
+      _clients--;
+    }
+  
+  if(_clients == 0)
+    {
+      NSLog(@"Client count is zero, exiting");
+      exit(0);
+    }
+  
+  NSLog(@"NSSpeechRecognizer server connection count = %d after disconnection", _clients);
+  while((o = [en nextObject]) != nil)
+    {
+      if ([o isKindOfClass: self])
+        {
+          RELEASE(o);
+        }
+    }
+}
+
++ (void)initialize
+{
+  _sharedInstance = [[self alloc] init];
+  _clients = 0;
+  [[NSNotificationCenter defaultCenter]
+		addObserver: self
+		   selector: @selector(connectionDied:)
+                       name: NSConnectionDidDieNotification
+                     object: nil];
+}
+
++ (void)start
+{
+  NSConnection *connection = [NSConnection defaultConnection];
+  [connection setRootObject: _sharedInstance];
+  RETAIN(connection);
+  if (NO == [connection registerName: @"GSSpeechRecognitionServer"])
+    {
+      NSLog(@"Could not register name, make sure another server is not running.");
+      return;
+    }
+  [[NSRunLoop currentRunLoop] run];
+}
+
++ (id)sharedServer
+{
+  NSLog(@"NSSpeechRecognizer server connection count = %d after connection", _clients);
+  return _sharedInstance;
+}
+
+- (void) addClient
+{
+  _clients++;
+}
+
+- (id)init
+{
+  if (nil == (self = [super init]))
+    {
+      return nil;
+    }
+  
+  _engine = [GSSpeechRecognitionEngine defaultSpeechRecognitionEngine];
+  if (nil == _engine)
+    {
+      [self release];
+      return nil;
+    }
+  else
+    {
+      NSLog(@"Got engine starting... %@", _engine);
+      [_engine start];
+    }
+
+  _blocking = [[NSMutableArray alloc] initWithCapacity: 10];  // 10 seems reasonable...
+
+  return self;
+}
+
+- (void) dealloc
+{
+  [_engine stop];
+  RELEASE(_engine);
+  RELEASE(_blocking);
+  [super dealloc];
+}
+
+- (void) addToBlockingRecognizers: (NSString *)s
+{
+  [_blocking addObject: s];
+}
+
+- (void) removeFromBlockingRecognizers: (NSString *)s
+{
+  [_blocking removeObject: s];
+}
+
+- (BOOL) isBlocking: (NSString *)s
+{
+  return [[_blocking firstObject] isEqualToString: s];
```

Comment:
You add to the end of the array, this means your stack is the other way around.You need to check the last element. Also you should return false if the array is empty.